ign_ros2_control/src/MimicJointSystem.cc

```diff
@@ -0,0 +1,402 @@
+/*!\file
+ *
+ * \author  Lovro Ivanov lovro.ivanov@gmail.com
+ * \date    2023-03-15
+ *
+ */
+//----------------------------------------------------------------------
+
+#include "ign_ros2_control/MimicJointSystem.hh"
+
+#include <string>
+#include <unordered_set>
+
+#include <ignition/common/Profiler.hh>
+#include <ignition/math/PID.hh>
+#include <ignition/plugin/Register.hh>
+
+#include "ignition/gazebo/components/JointForceCmd.hh"
+#include "ignition/gazebo/components/JointVelocityCmd.hh"
+#include "ignition/gazebo/components/JointPosition.hh"
+#include "ignition/gazebo/components/Joint.hh"
+#include "ignition/gazebo/components/Name.hh"
+#include "ignition/gazebo/components/JointType.hh"
+
+#include "ignition/gazebo/Model.hh"
+
+
+class ign_ros2_control::MimicJointSystemPrivate
+{
+/// \brief Joint Entity
+
+public:
+  ignition::gazebo::Entity jointEntity {ignition::gazebo::kNullEntity};
+
+public:
+  ignition::gazebo::Entity mimicJointEntity {ignition::gazebo::kNullEntity};
+
+/// \brief Joint name
+
+public:
+  std::string jointName;
+
+public:
+  std::string mimicJointName;
+
+/// \brief Commanded joint position
+
+public:
+  double jointPosCmd {0.0};
+
+public:
+  double multiplier {1.0};
+
+public:
+  double offset {0.0};
+
+  /// \brief Model interface
+
+public:
+  ignition::gazebo::Model model {ignition::gazebo::kNullEntity};
+
+  /// \brief Position PID controller.
+
+public:
+  ignition::math::PID posPid;
+
+  /// \brief Joint index to be used.
+
+public:
+  unsigned int jointIndex = 0u;
+
+public:
+  unsigned int mimicJointIndex = 0u;
+
+public:
+  double deadZone = 1e-6;
+
+  // approach adopted from https://github.com/gazebosim/gz-sim/blob/330eaf2f135301e90096fe91897f052cdaa77013/src/systems/joint_position_controller/JointPositionController.cc#L69-L77
+  /// \brief Operation modes
+  enum OperationMode
+  {
+    /// \brief Use PID to achieve positional control
+    PID,
+    /// \brief Bypass PID completely. This means the joint will move to that
+    /// position bypassing the physics engine.
+    ABS
+  };
+
+  /// \brief Joint position mode
+
+public:
+  OperationMode mode = OperationMode::PID;
+};
+
+namespace ign_ros2_control
+{
+
+  MimicJointSystem::MimicJointSystem()
+  : dataPtr(std::make_unique < ign_ros2_control::MimicJointSystemPrivate > ())
+  {
+  }
+
+  void MimicJointSystem::Configure(
+    const ignition::gazebo::Entity & _entity,
+    const std::shared_ptr < const sdf::Element > & _sdf,
+    ignition::gazebo::EntityComponentManager & _ecm,
+    ignition::gazebo::EventManager & /* _eventMgr*/)
+  {
+    // Make sure the controller is attached to a valid model
+    this->dataPtr->model = ignition::gazebo::Model(_entity);
+    if (!this->dataPtr->model.Valid(_ecm)) {
+      ignerr << "MimicJointSystem  Failed to initialize because [" <<
+        this->dataPtr->model.Name(_ecm)
+             << "] (Entity=" << _entity << ")] is not a model.. ";
```

Comment:
```suggestion
      ignerr << "MimicJointSystem  Failed to initialize because [" 
             << this->dataPtr->model.Name(_ecm)
             << "] (Entity=" << _entity << ")] is not a model.. ";
```